src/Aspire.Hosting/Eventing/DistributedApplicationEventing.cs

```diff
@@ -0,0 +1,41 @@
+namespace Aspire.Hosting.Eventing;
+
+internal class DistributedApplicationEventing : IDistributedApplicationEventing
+{
+    private readonly Dictionary<Type, List<DistributedApplicationEventSubscription>> _subscriptions = new();
+
+    public async Task PublishAsync<T>(T @event, CancellationToken cancellationToken) where T : IDistributedApplicationEvent
+    {
+        if (_subscriptions.TryGetValue(typeof(T), out var subscriptions))
+        {
+            foreach (var subscription in subscriptions)
+            {
+                await subscription.Callback(@event, cancellationToken).ConfigureAwait(false);
+            }
+        }
+    }
+
+    public DistributedApplicationEventSubscription Subscribe<T>(Func<T, CancellationToken, Task> callback) where T : IDistributedApplicationEvent
+    {
+        var subscription = new DistributedApplicationEventSubscription(async (@event, ct) =>
```

Comment:
Should we synchronize access to subscription list in case this is called concurrently?